File: android/src/main/kotlin/fr/pointcheval/native_crypto/HashAlgorithm.kt
package fr.pointcheval.native_crypto

import java.security.MessageDigest

enum class HashAlgorithm(val length : Int) {
    SHA1(160),
    SHA128(128),
    SHA256(256),
    SHA512(512);
}


class Hash() {
    fun digest(data: ByteArray?, algorithm: HashAlgorithm): ByteArray {
        val func : String = when (algorithm) {
            HashAlgorithm.SHA1 -> "SHA-1"
            HashAlgorithm.SHA128 -> "SHA-128"
            HashAlgorithm.SHA256 -> "SHA-256"
            HashAlgorithm.SHA512 -> "SHA-512"
        }
        val md = MessageDigest.getInstance(func)
        return md.digest(data)
    }

    fun digest(data: ByteArray?, algorithm: String): ByteArray {
        val func : HashAlgorithm = when (algorithm) {
            "sha1" -> HashAlgorithm.SHA1
            "sha128" -> HashAlgorithm.SHA128
            "sha256" -> HashAlgorithm.SHA256
            "sha512" -> HashAlgorithm.SHA512
            else -> HashAlgorithm.SHA256
        }
        return digest(data, func)
    }

    fun digest(data: ByteArray?): ByteArray {
        return digest(data, HashAlgorithm.SHA256)
    }
}

File: packages/native_crypto/lib/src/utils/hash_algorithm.dart
// Author: Hugo Pointcheval
// Email: git@pcl.ovh
// -----
// File: hash_algorithm.dart
// Created Date: 23/05/2022 22:01:59
// Last Modified: 26/05/2022 22:59:04
// -----
// Copyright (c) 2022

import 'dart:typed_data';

import 'package:native_crypto/src/platform.dart';
import 'package:native_crypto/src/utils/extensions.dart';
import 'package:native_crypto_platform_interface/native_crypto_platform_interface.dart';

/// Defines the hash algorithms.
enum HashAlgorithm {
  sha256,
  sha384,
  sha512;

  /// Digest the [data] using this [HashAlgorithm].
  Future<Uint8List> digest(Uint8List data) async {
    Uint8List? hash;
    try {
      hash = await platform.digest(data, name);
    } catch (e, s) {
      throw NativeCryptoException(
        message: '$e',
        code: NativeCryptoExceptionCode.platform_throws.code,
        stackTrace: s,
      );
    }

    if (hash.isNull) {
      throw NativeCryptoException(
        message: 'Failed to digest data! Platform returned null.',
        code: NativeCryptoExceptionCode.platform_returned_null.code,
      );
    }

    if (hash!.isEmpty) {
      throw NativeCryptoException(
        message: 'Failed to digest data! Platform returned no data.',
        code: NativeCryptoExceptionCode.platform_returned_empty_data.code,
      );
    }

    return hash;
  }
}
